lib/puppet/type/agent_upgrade.rb

```diff
@@ -0,0 +1,30 @@
+Puppet::Type.newtype(:agent_upgrade) do
+  ensurable
+
+  newparam(:name, :namevar => true)
+
+  newparam(:source) do
+    newvalues /^((https?|puppet):\/\/)|\\\\\S+\\\w+|[a-z]:\\/i
+    validate do |value|
+      if value !~ /^((https?|puppet):\/\/)|\\\\\S+\\\w+|[a-z]:\\/i
+        fail "Please provide a valid http(s), puppet or unc path"
+      end
+      if value =~ /^https?/ && value !~ URI::regexp
+        fail("Please provide a valid URI")
+      end
+    end
+  end
+
+  newparam(:arch) do
+    newvalues /^(x86|x64)$/
```

Comment:
I see you did the following in the provider - Facter.value(:architecture)

I would ensure that you mention that here in the description of this you are going to write.